### DISCUSSION TITLE
@lit-labs/ssr - Property binding server side?

### DISCUSSION THREAD

**Discussion by pedrolzoliveira** [lit/lit#4406] [Q&A] [answered]
Hello everyone. I'm currently using `@lit-labs/ssr` in one of my projects and I'm facing an issue when trying to pass props to a `LitElement` rendered on the server side.

It works fine with primitive types, but when I try to pass an object or array to the element, it breaks.

I wrote some examples.

Here's the code for the element, nothing fancy, only a simple LitElement with 2 properties, it renders the name and a list with the items.
```js
import { LitElement, html } from "lit";
import { repeat } from "lit/directives/repeat.js";

export class MyListElement extends LitElement {
  static properties = {
    items: { type: Array },
    name: { type: String }
  }

  constructor() {
    super();
    this.name = 'Tony Stark';
    this.items = ['Mark 42', 'Mark 43'];
  }
  
  render() {
    return html`
      <div>
        <h1>${this.name}'s list</h1>
        <ul>
          ${
            repeat(
              this.items,
              (item) => html`<li>${item}</li>`
            )
          }
        </ul>
      </div>`;
  }
}

customElements.define('my-list', MyListElement);

```

Here's the server rendering the examples
```js
import express from 'express';
import { html } from 'lit';
import { render } from '@lit-labs/ssr/lib/render.js';
import { RenderResultReadable } from '@lit-labs/ssr/lib/render-result-readable.js';

const server = express();

server.use('/dist', express.static('dist'));

server.use((req, res, next) => {
  res.renderTemplate = (template) => {
    res.setHeader('Content-Type', 'text/html; charset=utf-8');
    const readableResult = new RenderResultReadable(render(template));
    readableResult.pipe(res);
  }
  next();
})

const myTemplate = html`
  <html>
    <head>
      <script src="/dist/bundle.js"></script>
    </head>
    <body>
      <h1>The list</h1>

      <!-- Renders name correctly. -->
      <my-list name=${'Example 1'}></my-list>

      <!-- Does not renders the name correctly. -->
      <my-list .name=${'Example 2'}></my-list>

      <!-- Does not renders the element at all. -->
      <my-list items=${['Example 3']}></my-list>

      <!-- Does not render the items correctly. -->
      <my-list .items=${['Example 4']}></my-list>

    </body>
  </html>
`;

server.get('/', (req, res) => {
  res.renderTemplate(myTemplate);
});

server.listen(3000, () => {
  console.log('Server is listening on port 3000');
});
```
Here's a git repo with the examples

https://github.com/pedrolzoliveira/lit-property-binding-ssr/

It seems logical for me that the third example fails, because that's not how you should pass props to the elements in Lit.

For the rest looks like it might be a problem with property bindings? Are we supposed to use it that way when rendering on the server?

I'm looking for a way to make it works for some weeks now, in my project I was able to do that by converting the object/array to a string and them parsing it back to object/array.

**Comment by AndrewJakubowicz** [comment, marked as answer]
Hi! Thank you for asking this question. It definitely highlights a lack of good errors when hitting painful cases in Lit SSR.
I've attached to this comment the changes required to fix the issues you are seeing (found at the end).

### Issue 1: Your client side elements load too fast

`<script src='/dist/bundle.js'></script>` is located at the top of the HTML file, which blocks the HTML parser to fetch and execute your JavaScript. This results in `my-list` being completely loaded and ready to use before the HTML parser has started to render the body element.

This means that if you have server side rendered `my-list`, the declarative shadow DOM will have already attached the shadowRoot and an error gets thrown in the browser: `Shadow root cannot be created on a host which already hosts a shadow tree.`

The solution to this is to not load your component until the initial render of the page is complete. Either lazy-load your component within the script, (e.g. with a dynamic import). Or add `defer` to the script tag. Or move it to the bottom (as I have done).

### Issue 2: server.js is not server side rendering `my-list`.

Your `server.js` code isn't server side rendering the LitElement `<my-list>`. When running your example, if you inspect the network tab to see what `express` is serving as the initial HTML render, you'll notice that the HTML contains `<my-list></my-list>`. This is the biggest clue as to why the *Example 1* `attribute` case works. Example 1 works because the server sends down: `<my-list name="Example 1"></my-list>`. And because your custom element is already defined on the client, it uses that attribute to populate it's `name`.

Example 2 doesn't work, because the server isn't server side rendering the initial render of `<my-list .name=${'Example 2'}></my-list>`. So instead `<my-list></my-list>` is being sent in the HTML, and when the element upgrades on the client, it has lost what state it had.

To server side render `<my-list>`, and get the initial render to include the state on the server, add an import to `import './elements/my-list.js';` in `server.js`.
This lets `@lit-labs/ssr` find the element on the server to generate the server side HTML.

After adding the import to `server.js`, the initial HTML in the network tab contains:

```html
<!--Before adding the server.js import:-->
<my-list name="Example 1"></my-list>
```

```html
<!--After adding the server.js import:-->
<my-list  name="Example 1"><template shadowroot="open" shadowrootmode="open"><!--lit-part v5WoVya4F4E=-->
      <div>
        <h1><!--lit-part-->Example 1<!--/lit-part-->'s list</h1>
        <ul>
          <!--lit-part--><!--lit-part PuAyB/F8cHw=--><li><!--lit-part-->Mark 42<!--/lit-part--></li><!--/lit-part--><!--lit-part PuAyB/F8cHw=--><li><!--lit-part-->Mark 43<!--/lit-part--></li><!--/lit-part--><!--/lit-part-->
        </ul>
      </div><!--/lit-part--></template></my-list>
```

Now `@lit-labs/ssr` is server side rendering your components, so on the initial render (and without JavaScript enabled), the elements look correct, and contain the properties that are expected.

I'm going to file some bugs from this investigation.
 - I need to check that [lit.dev/ ssr documentation](https://lit.dev/docs/ssr/overview/) mentions that the LitElements need to be loaded on the server.
 - I'll file a bug to figure out if using a property binding on something that isn't SSR'd might need to raise a helpful error. That information doesn't make it to the client. (There is subtly here, because you may want the property binding to be used on the client. For example `<input .value=${...}>` doesn't have an SSR'd representation, but is useful on the client. Another alternative is to use [`server only templates`](https://www.npmjs.com/package/@lit-labs/ssr#server-only-templates) on the server.
 - Defining your LitElement before HTML parse should give a more helpful error if the LitElement encounters an existing shadowRoot.


### Code that fixes issues:

```diff
--- a/src/elements/my-list.js
+++ b/src/elements/my-list.js
@@ -1,3 +1,9 @@
+// Wherever your entry file in to the app is. If you want to support SSR of Lit
+// elements, you must first install lit-element-hydrate-support. This lets
+// LitElements's load from SSR'd content. Reference:
+// https://lit.dev/docs/ssr/client-usage/#loading-@lit-labsssr-clientlit-element-hydrate-support.js
+import '@lit-labs/ssr-client/lit-element-hydrate-support.js';
+
 import { LitElement, html } from "lit";
 import { repeat } from "lit/directives/repeat.js";

``` 

```diff
--- a/src/server.js
+++ b/src/server.js
@@ -2,6 +2,7 @@ import express from 'express';
 import { html } from 'lit';
 import { render } from '@lit-labs/ssr/lib/render.js';
 import { RenderResultReadable } from '@lit-labs/ssr/lib/render-result-readable.js';
+import './elements/my-list.js'; // <- Need to load element to SSR in the server.
 
 const server = express();
 
@@ -19,7 +20,13 @@ server.use((req, res, next) => {
 const myTemplate = html`
   <html>
     <head>
-      <script src="/dist/bundle.js"></script>
+      <!-- 
+        By loading my-list here, it's already fully defined and upgraded by the
+        time the html is being parsed. This removes any need for SSR, but also
+        creates an error if the browser encounters the SSR'd element later.
+        Moving loading script to the end resolves this issue – and also captures
+        the intent of SSR more closely.
+      -->
     </head>
     <body>
       <h1>The list</h1>
@@ -31,11 +38,12 @@ const myTemplate = html`
       <my-list .name=${'Example 2'}></my-list>
 
       <!-- Does not renders the element at all. -->
-      <my-list items=${['Example 3']}></my-list>
+      <!-- Edit: Removed because example 3 is invalid and causes SSR to throw. -->
 
       <!-- Does not render the items correctly. -->
       <my-list .items=${['Example 4']}></my-list>
 
+      <script src='/dist/bundle.js'></script>
     </body>
   </html>
 `;
```

**Reply by AndrewJakubowicz** [reply]
`@lit-labs/ssr` is a toolbox that lets you create your own server side rendering -> client side rendering strategies. For your case, I'd recommend the following pattern.

On the server use **server only templates**. These templates prevent binding to properties or events, and ensure your LitElement's are SSR'd with data that is also sent down to the client.
See https://stackblitz.com/edit/github-r3xs9z?file=src%2Fserver.js

**Reply by pedrolzoliveira** [reply]
Thank you for the reply! I was doing it wrong for some time now lol

really appreciate the work you guys being putting in lit

**Comment by augustjk** [comment]
In addition to @AndrewJakubowicz 's detailed explanation of the issues, I do want to point out that property bindings on custom elements in regular `html` tag templates do work. The element renderer will set those properties before generating the declarative shadow DOM. It's just that HTML has no way of setting properties on elements so the emitted string won't contain them. This makes hydration a little tricky as the markup doesn't contain all the data needed for hydration.

The SSR render method allows for an option called `deferHydration` which prevents Lit elements from doing their first update by adding a `defer-hydration` attribute to the element. This gives an opportunity to imperative set the needed properties, then remove the `defer-hydration` attribute to initiate hydration and first update.
See an example here: https://stackblitz.com/edit/github-r3xs9z-cyxzog?file=src%2Fserver.js

This definitely is more clunky and might be better with a more framework like layer of abstraction which we currently don't have. So if you're able to serialize into attributes, that might be the easiest thing to do.

**Reply by pedrolzoliveira** [reply]
Thanks for the information, that's going to be useful for me.

thank you guys for the work in lit!
